scipy/special/cephes/ellpj.c

```diff
@@ -93,17 +100,44 @@ int ellpj(double u, double m, double *sn, double *cn, double *dn, double *ph)
         *dn = 1.0 - 0.5 * m * t * t;
         return (0);
     }
+
+    /* if m is near 1, use approximation for amplitude
+       given in Abramowitz and Stegun, 16.15.4 */
     if (m >= 0.9999999999) {
-        ai = 0.25 * (1.0 - m);
-        b = cosh(u);
-        t = tanh(u);
-        phi = 1.0 / b;
-        twon = b * sinh(u);
-        *sn = t + ai * (twon - u) / (b * b);
-        *ph = 2.0 * atan(exp(u)) - NPY_PI_2 + ai * (twon - u) / b;
-        ai *= t * phi;
-        *cn = phi - ai * (twon - u);
-        *dn = phi + ai * (twon + u);
+        /* Evaluate the complete elliptic integral using the
+           1-m<<1 approximation in ellpk.c */
+        K = ellpk(1.0-m);
+        uabs = fabs(u);
+        j = (int)(uabs/K);
+        du = uabs-((double)j)*K;
+        r = j%4;
+
+        /* When phi is in the I and III quadrants */
+        if(r == 0 || r == 2) {
+            t = cosh(du);
+            phi = 0.25*(1.0-m)*(sinh(du)-du/t);
+            /* add Gudermannian term */
+            phi += 2.0*atan(exp(du)) - NPY_PI_2;
+        }
+        /* When phi is in the II and IV quadrants */
+        else{
+            du = K-du;
+            t = cosh(du);
+            phi = 0.25*(1.0-m)*(sinh(du)-du/t);
+            /* add Gudermannian term */
+            phi += 2.0*atan(exp(du)) - NPY_PI_2;
+            phi = NPY_PI_2-phi;
+        }
```

Comment:
To avoid redundancy, perhaps:
```suggestion
        /* When phi is in the II/IV quadrants, work with complements */
        if(r == 1 || r == 3) {
            du = K-du;
        }
        
        /* A&S 16.15.4 */
        m1 = 1.0 - m
        gd = 2.0*atan(exp(du)) - NPY_PI_2;
        phi = gd + 0.25*m1*(sinh(du) - du/cosh(du));

        if(r == 1 || r == 3) {
            phi = NPY_PI_2-phi;
        }
```
(if we can tolerate the allocation of a few extra floats)